pybart/new_matcher.py

```diff
@@ -0,0 +1,330 @@
+    
+    # return the set of captured labels between the two tokens, given their indices
+    def edge(self, t1: int, t2: int) -> Set[str]:
+        return self.indices2label.get((t1, t2), default=None)  # TODO - is it legitimate to return None here?
+
+
+class GlobalMatcher:
+    def __init__(self, constraint: Full):
+        self.constraint = constraint
+        self.captured_labels = defaultdict(set)
+        # list of token ids that don't require a capture
+        self.dont_capture_names = [token.id for token in constraint.tokens if not token.capture]
+    
+    # filter a single match group according to non-token constraints
+    def _filter(self, match: Mapping[str, int], sentence) -> bool:
+        # check that the distance between the tokens is not more than or exactly as required
+        for distance in self.constraint.distances:
+            if distance.token1 not in match or distance.token2 not in match:
+                continue
+            calculated_distance = sentence.index(match[distance.token2]) - sentence.index(match[distance.token1]) - 1
+            # TODO - maybe move the following switch case into a method of the Distance constraint? will simplify this code to
+            #   "if not distance.satisfied(calculated_distance): return False"
+            if isinstance(distance, ExactDistance) and distance.distance != calculated_distance:
+                return False
+            elif isinstance(distance, UptoDistance) and distance.distance < calculated_distance:
+                return False
+            else:
+                raise ValueError("Unknown Distance type")  # TODO - change to our exception?
+        
+        # check for a two-word or three-word phrase match in a given phrases list
+        for concat in self.constraint.concats:
+            # TODO -
+            #   1. again, maybe should be moved to constraint's logic
+            #   2. can it be partially moved to initialization time (optimization)?
+            if isinstance(concat, TokenPair):
+                if concat.token1 not in match or concat.token2 not in match:
+                    continue
+                word_indices = [match[concat.token1], match[concat.token2]]
+            elif isinstance(concat, TokenTriplet):
+                if concat.token1 not in match or concat.token2 not in match or concat.token3 not in match:
+                    continue
+                word_indices = [match[concat.token1], match[concat.token2], match[concat.token3]]
+            else:
+                raise ValueError("Unknown TokenTuple type")  # TODO - change to our exception?
+            if "_".join(sentence.get_text(w) for w in word_indices) not in concat.tuple_set:
+                return False
+        
+        return True
+        
+    @staticmethod
+    def _try_merge(base_assignment: Mapping[str, int], new_assignment: Mapping[str, int]) -> Mapping[str, int]:
```

Comment:
i didn't look closely into this code. i trust that you tested that it works? may be wise to add some test-cases. (for the entire algorithm, not just for _try_merge)

Reply:
@yoavg, should I open a new/separate issue for testing the new_matcher, and resolve this comment?